### PR TITLE
ci: reproducible builds ko (no gcp)

### PR DESCRIPTION
**Pull request by leongross**
### Proposed change(s)

- Reproducible container images using `ko` for our microservices:
  - Node Operator
  - Join Service
  - KMS
  - Verification Service
- Base Image building using `apko`
  - Generate and upload `apko` images and signatues
  - Generate and upload SBOMS and signatures
  

### Checklist

- [ ] Update [docs](https://github.com/edgelesssys/constellation/tree/main/docs)
- [x] Add labels (e.g., for changelog category)
- [x] Link to Milestone

### CODE REVIEW

**Comment by leongross** [comment]
E2E: https://github.com/edgelesssys/constellation/commit/0f00716f6f8c804276c5b15d100a6bc2cce9d842

File: .github/actions/build_apko/action.yml
Comment (datosh) [inline .github/actions/build_apko/action.yml:16]: Maybe this limitation should be stated in the description? It's the first thing I read when I use an Action I have not used before.

File: .github/workflows/build-apko-image.yml
Comment (malt3) [inline .github/workflows/build-apko-image.yml:9]: I think this could be allowed on any branch. Some devs build microservices on their feature branch during testing. The same applied to all workflows.

File: .github/workflows/build-constellation-node-operator-ko.yml
Comment (malt3) [inline .github/workflows/build-constellation-node-operator-ko.yml:27]: This can be skipped when building with ko

**Comment by malt3** [comment]
General question: Should those workflows replace the existing workflows? Having duplicate workflows sounds hard to maintain.

File: .github/actions/build_apko/action.yml
Comment (datosh) [inline .github/actions/build_apko/action.yml:28]: Dependent actions should be referenced by hash with version as comment.

```suggestion
      uses: redhat-actions/podman-login@7283c5d4ffb37e87a455332f0d68b144536d842f # v1.5
```

File: .github/actions/build_micro_service_ko/action.yml
Comment (datosh) [inline .github/actions/build_micro_service_ko/action.yml:44]: Probably left over debugging code? 

```suggestion
      if: ${{ inputs.pseudoVersion == 'true' }}
```

File: .github/workflows/build-apko-image.yml
Comment (datosh) [inline .github/workflows/build-apko-image.yml:9]: before merging, same for other workflow files
```suggestion
```

File: .github/actions/build_apko/action.yml
Comment (datosh) [inline .github/actions/build_apko/action.yml:60]: We should also use cosign for the `apko` generated images.

`libcryptsetup` will be the final image, right? So users should be able to check it.

In case where we use `apko` images as a base image for further steps, we should check the signature before proceeding to use it.

**Review by datosh — COMMENTED** [review]
Looks good to me so far!

Is there already a plan how to switch from the current build code to this one, since this PR removes none of the build code for containers we have right now? I don't think we should have / maintain two build paths on `main`, or do we require more testing before we are comfortable removing the current one?

File: .github/workflows/build-apko-image.yml
Comment (leongross) [reply .github/workflows/build-apko-image.yml:9]: The initial ideas was to prevent unnecessary re-building without changes to the image definitions.
Since all the `apko` image definitions should reside in `image/apko` anyways, we could maybe remove the branch filtering and add path filtering for the said folder?

This could look like this:
```suggestion
          paths:
            - 'image/apko/*.yaml'

```

File: .github/actions/build_micro_service_ko/action.yml
Comment (leongross) [reply .github/actions/build_micro_service_ko/action.yml:44]: Yes, this is a debug relic and should be removed.

File: .github/workflows/build-apko-image.yml
Comment (leongross) [reply .github/workflows/build-apko-image.yml:9]: Yes, I added this branch filter to run the tests on this branch, this will be removed before merge.

Comment (malt3) [reply .github/workflows/build-apko-image.yml:9]: I slightly misread the code and thought the branch filter was on the `workflow_dispatch`. Adding an additional path filter for the apko yamls on this workflow sounds like a good idea though :)

File: .github/workflows/build-joinservice-image-ko.yml
Comment (katexochen) [inline .github/workflows/build-joinservice-image-ko.yml:29]:
```suggestion
          go-version: "1.19.4"
```

File: .github/workflows/build-kms-image-ko.yml
Comment (katexochen) [inline .github/workflows/build-kms-image-ko.yml:29]:
```suggestion
          go-version: "1.19.4"
```

File: .github/workflows/build-kms-image-ko.yml
Comment (katexochen) [inline .github/workflows/build-kms-image-ko.yml:32]: Pin hash.

File: .github/workflows/build-verification-service-ko.yml
Comment (katexochen) [inline .github/workflows/build-verification-service-ko.yml:27]:
```suggestion
          go-version: "1.19.4"
```

File: .ko.yaml
Comment (katexochen) [inline .ko.yaml:12]:
```suggestion
  - CGO_ENABLED=0
  - GOOS=linux
  - GOARCH=amd64
```
Is there a reason why kms and joinservice lack GOOS and GOARCH? Same for missing `-a `flag.

File: image/apko/alpine-base-user-65532.yaml
Comment (katexochen) [inline image/apko/alpine-base-user-65532.yaml:1]: nit: Imo image dir should contain OS image only, maybe create a separate top lvl dir named `container`, `apko`, etc.

File: .github/actions/build_apko/action.yml
Comment (daniel-weisse) [inline .github/actions/build_apko/action.yml:108]: indentation is off for the `podman run` part

File: .github/actions/build_apko/action.yml
Comment (leongross) [reply .github/actions/build_apko/action.yml:120]: Currently, `libcryptsetup` is only part of the [alpine-libcryptsetup.yaml](./image/apko/alpine-libcryptsetup.yaml) image that can be used as a base image for an `apko` container, if the application running in this container needs this libraray.

> I was trying to motivate adding cosign signatures to the `apko` images so that "users of the libcryptsetup image can check the signature on it via cosign verify".

I'm not sure if I understand what you mean. Which signatures should be stored in which images and in when should these signatures be verified?

File: .github/workflows/build-kms-image-ko.yml
Comment (katexochen) [reply .github/workflows/build-kms-image-ko.yml:29]:
```suggestion
          go-version: "1.19.5"
```

File: .github/actions/build_apko/action.yml
Comment (leongross) [reply .github/actions/build_apko/action.yml]: Due to issues with the `ghcr.io` registry I decided to roll back to docker, so I think this is obsolete.

File: .ko.yaml
Comment (leongross) [reply .ko.yaml:12]: No there is no reason for that. I think your idea is good to make it more explicit.

File: image/apko/alpine-base-user-65532.yaml
Comment (leongross) [reply image/apko/alpine-base-user-65532.yaml:1]: maybe we could have opinion of the code owner of the `images/` folder (@malt3) on that?

Comment (malt3) [reply image/apko/alpine-base-user-65532.yaml:1]: Probably the right call. I'd add a separate folder for this since they are used for very different purposes.

File: .github/actions/build_micro_service_ko/action.yml
Comment (malt3) [inline .github/actions/build_micro_service_ko/action.yml:43]:
```suggestion
```

File: .github/actions/build_ko/action.yml
Comment (malt3) [inline .github/actions/build_ko/action.yml:108]:
```suggestion
      uses: actions/upload-artifact@0b7f8abb1508181956e8e162db84b466c27e18ce # v3.1.2
```

**Review by malt3 — APPROVED** [review]
LGTM. Slight changes will be needed to fully switch over the release pipeline but I would be in favor of merging as it is.

File: .github/actions/build_apko/action.yml
Comment (katexochen) [inline .github/actions/build_apko/action.yml:167]: Why does this need to run even on error?

Comment (leongross) [reply .github/actions/build_apko/action.yml:167]: This seems to be a debugging artifact, it should be safe to remove.

**Pull request merged**
